stores/json/store.go

```diff
@@ -211,7 +211,9 @@ func (store Store) jsonFromTreeBranch(branch sops.TreeBranch) ([]byte, error) {
 
 func (store Store) treeBranchFromJSON(in []byte) (sops.TreeBranch, error) {
 	dec := json.NewDecoder(bytes.NewReader(in))
-	dec.Token()
+	if _, err := dec.Token(); err != nil {
+		return nil, err
```

Comment:
Could you please document what exactly this is doing? I guess this is "Tests for, but does not disallow, duplicate keys in json.", but it is totally unclear to me how this code achieves that.

Reply:
This change is (as far as I can tell, been a while) not required, it's just a drive-by change that I made: `Token()` does return an error, and checking+returning it is correct. 
Later on, the `store.treeBranchFromJSONDecoder(dec)` call will internally do the same thing
```
func (store Store) treeItemFromJSONDecoder(dec *json.Decoder) (sops.TreeItem, error) {
	var item sops.TreeItem
	key, err := dec.Token()
	if err != nil && err != io.EOF {
		return item, err
	}
```
So I guess I figured that we should do it on the very first instance too

Reply:
Except that this one fails when `err == io.EOF`. Which I guess is OK. I would add a comment explaining that.

Reply:
Happy to add a comment, but I'm not sure what to write: I don't quite see why we need to document why we handle an error. Usually, motivation/explanation is required as for why one omits to handle an error.

Reply:
After dwelling in sources of the JSON module, I now understand why this change is correct.

The main problem I have with this change is that it takes a long time to figure out what it does, and since it is totally unrelated to the commit message it appears in it is very confusing to figure out where it comes from.

It would be a lot better to move this to a separate commit (or even a separate PR) with a clear commit message (something like "Make sure errors returned when extracting the first token are handled").

Also there is another potential error lurking here: the first return value also needs to be checked whether it's a `{`. Otherwise `LoadPlainFile` produces "strange" error messages that aren't very helpful:
```console
$ echo '[]' | sops --encrypt --input-type json /dev/stdin
Error unmarshalling file: Could not unmarshal input data: Expected JSON object key, got ] of type json.Delim instead
$ echo '[}' | sops --encrypt --input-type json /dev/stdin
Error unmarshalling file: Could not unmarshal input data: invalid character '}' looking for beginning of value
$ echo '["foo","bar"]' | sops --encrypt --input-type json /dev/stdin
Error unmarshalling file: Could not unmarshal input data: invalid character '}' after array element
$ echo '["foo":"bar"}' | sops --encrypt --input-type json /dev/stdin
Error unmarshalling file: Could not unmarshal input data: invalid character ':' after array element
```

Reply:
I don't see what difference this makes. As far as I can tell this PR does not change the behaviour against master

This PR:
```
$  echo '[]' | go run ./cmd/sops/ --encrypt --input-type json /dev/stdin
Error unmarshalling file: Could not unmarshal input data: Expected JSON object key, got ] of type json.Delim instead
$  echo '[}' | go run ./cmd/sops/ --encrypt --input-type json /dev/stdin
Error unmarshalling file: Could not unmarshal input data: invalid character '}' looking for beginning of value
$  echo '["foo","bar"]' | go run ./cmd/sops/ --encrypt --input-type json /dev/stdin
Error unmarshalling file: Could not unmarshal input data: Expected JSON object key, got ] of type json.Delim instead
$  echo '["foo":"bar"}' | go run ./cmd/sops/ --encrypt --input-type json /dev/stdin
Error unmarshalling file: Could not unmarshal input data: invalid character ':' after array element
```


Master:
```
$  echo '[]' | go run ./cmd/sops/ --encrypt --input-type json /dev/stdin
Error unmarshalling file: Could not unmarshal input data: Expected JSON object key, got ] of type json.Delim instead
$  echo '[}' | go run ./cmd/sops/ --encrypt --input-type json /dev/stdin
Error unmarshalling file: Could not unmarshal input data: invalid character '}' looking for beginning of value
$ echo '["foo","bar"]' | go run ./cmd/sops/ --encrypt --input-type json /dev/stdin
Error unmarshalling file: Could not unmarshal input data: Expected JSON object key, got ] of type json.Delim instead
$ echo '["foo":"bar"}' | go run ./cmd/sops/ --encrypt --input-type json /dev/stdin
Error unmarshalling file: Could not unmarshal input data: invalid character ':' after array element
```
But sure I can remove this change if you want

Reply:
> Also there is another potential error lurking here: the first return value also needs to be checked whether it's a {.

It should not literally check that it's a `{`, since whitespace is also accepted. I think the error messages are pretty good. 

Consider `["foo":"bar"}`. 
- If we say `error: it must start with {`, then I think it's mixing different layers. 

1. First of all, the input needs to be valid `json`. The value `["foo":"bar"}` is not valid `json`, because a json list cannot contain key-value pairs, so the parse fails. The user will have to fix up his json, there are many tools to validate / lint `json`.
2. Secondly, there are app-specific constraints on what is required of the json object: such as that it is a dict, not a list. But those come later, IMO.

Reply:
> It should not literally check that it's a `{`, since whitespace is also accepted. I think the error messages are pretty good.

Yes, it should for a literal `{`, since whitespace has already been processed by the decoder.

> First of all, the input needs to be valid `json`. The value `["foo":"bar"}` is not valid `json`, because a json list cannot contain key-value pairs, so the parse fails. The user will have to fix up his json, there are many tools to validate / lint json.

True, but did you also consider the `[]` example? Or tried out something like `[{}]`? The error messages are not really helpful, because the code already mixes the two layers: it parses valid JSON, and at the same time only parses JSON objects.

> Secondly, there are app-specific constraints on what is required of the json object: such as that it is a dict, not a list. But those come later, IMO.

The function is called `treeBranchFromJSON`. It cannot process anything else than a JSON object.

In any case, this discusison IMO shows why this should not part of a totally unrelated commit. This has nothing to do with (not) checking for duplicate keys.